File: app/lib/models/ingredient.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'ingredient.freezed.dart';
part 'ingredient.g.dart';

@freezed
abstract class Ingredient with _$Ingredient {
  const factory Ingredient({
    required String id,
    required String name,
    @Default('') String description,
    @JsonKey(name: 'image_path') @Default('') String imagePath,
    @JsonKey(name: 'unit_id') required String unitId,
    @JsonKey(name: 'current_stock') @Default(0) double currentStock,
    @JsonKey(name: 'current_cost_per_unit') @Default(0) double currentCostPerUnit,
    @JsonKey(name: 'low_stock_threshold') double? lowStockThreshold,
    @JsonKey(name: 'price_alert_percentage') @Default(10) double priceAlertPercentage,
    @JsonKey(name: 'category_id') String? categoryId,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
    IngredientUnit? unit,
    IngredientCategory? category,
  }) = _Ingredient;

  factory Ingredient.fromJson(Map<String, dynamic> json) =>
      _$IngredientFromJson(json);
}

@freezed
abstract class IngredientCategory with _$IngredientCategory {
  const factory IngredientCategory({
    required String id,
    required String name,
    @JsonKey(name: 'sort_order') @Default(0) int sortOrder,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _IngredientCategory;

  factory IngredientCategory.fromJson(Map<String, dynamic> json) =>
      _$IngredientCategoryFromJson(json);
}

@freezed
abstract class IngredientUnit with _$IngredientUnit {
  const factory IngredientUnit({
    required String id,
    required String name,
    required String abbreviation,
    @JsonKey(name: 'unit_type') required String unitType,
  }) = _IngredientUnit;

  factory IngredientUnit.fromJson(Map<String, dynamic> json) =>
      _$IngredientUnitFromJson(json);
}

@freezed
abstract class PriceHistory with _$PriceHistory {
  const factory PriceHistory({
    required String id,
    @JsonKey(name: 'ingredient_id') required String ingredientId,
    @JsonKey(name: 'old_cost_per_unit') required double oldCostPerUnit,
    @JsonKey(name: 'new_cost_per_unit') required double newCostPerUnit,
    @JsonKey(name: 'change_percentage') required double changePercentage,
    required String source,
    @JsonKey(name: 'recorded_at') DateTime? recordedAt,
  }) = _PriceHistory;

  factory PriceHistory.fromJson(Map<String, dynamic> json) =>
      _$PriceHistoryFromJson(json);
}

File: internal/model/ingredient.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Unit struct {
	ID               uuid.UUID `json:"id"`
	Name             string    `json:"name"`
	Abbreviation     string    `json:"abbreviation"`
	UnitType         string    `json:"unit_type"`
	BaseUnitID       *uuid.UUID `json:"base_unit_id"`
	ConversionFactor *float64  `json:"conversion_factor"`
	CreatedAt        time.Time `json:"created_at"`
}

type IngredientCategory struct {
	ID        uuid.UUID `json:"id"`
	Name      string    `json:"name"`
	SortOrder int       `json:"sort_order"`
	IsActive  bool      `json:"is_active"`
	CreatedAt time.Time `json:"created_at"`
}

type Ingredient struct {
	ID                   uuid.UUID  `json:"id"`
	Name                 string     `json:"name"`
	Description          string     `json:"description"`
	ImagePath            string     `json:"image_path"`
	UnitID               uuid.UUID  `json:"unit_id"`
	CurrentStock         float64    `json:"current_stock"`
	CurrentCostPerUnit   float64    `json:"current_cost_per_unit"`
	LowStockThreshold    *float64   `json:"low_stock_threshold"`
	PriceAlertPercentage float64    `json:"price_alert_percentage"`
	CategoryID           *uuid.UUID `json:"category_id"`
	IsActive             bool       `json:"is_active"`
	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`

	// Persisted bulk-purchase defaults (entered via the purchase calculator).
	// Used as fallback when no mapped vendor_bill_items row exists.
	PurchaseQty     *float64   `json:"purchase_qty,omitempty"`
	PurchaseUnitID  *uuid.UUID `json:"purchase_unit_id,omitempty"`
	PurchasePrice   *float64   `json:"purchase_price,omitempty"`

	// Joined fields
	Unit        *Unit               `json:"unit,omitempty"`
	Category    *IngredientCategory `json:"category,omitempty"`
	LastPriceAt *time.Time          `json:"last_price_at,omitempty"`

	// Effective bulk price for list display: either the latest mapped vendor
	// bill item, or the saved purchase defaults above. Populated only by List.
	BulkQty       *float64 `json:"bulk_qty,omitempty"`
	BulkUnitAbbr  string   `json:"bulk_unit_abbr,omitempty"`
	BulkTotal     *float64 `json:"bulk_total,omitempty"`
}

type IngredientPriceHistory struct {
	ID               uuid.UUID `json:"id"`
	IngredientID     uuid.UUID `json:"ingredient_id"`
	OldCostPerUnit   float64   `json:"old_cost_per_unit"`
	NewCostPerUnit   float64   `json:"new_cost_per_unit"`
	ChangePercentage float64   `json:"change_percentage"`
	Source           string    `json:"source"`
	BillID           *uuid.UUID `json:"bill_id"`
	RecordedAt       time.Time `json:"recorded_at"`
}
